python/metatensor-operations/metatensor/operations/to.py

```diff
@@ -116,19 +120,81 @@ def block_to(
                         " gradient tracking"
                     )
 
-    return _block_to(block, backend, dtype, device, requires_grad)
+    # Walk the tree of gradients without recursion
+    # (recursion is not supported by torchscript)
+    current_location: List[str] = []
+    current_block = block
+    transformed_blocks: List[TensorBlock] = []
+    last_visited = ""
+
+    while True:
```

Comment:
What we do here is a depth-first search. Maybe it is worth if there a pseudocode which is a bit simpler. There is for example one here:

https://en.wikipedia.org/wiki/Depth-first_search

Also it maybe make sense to extract the code in the end to made this available for other classes.

Reply:
I've read and understood the algorithm. My first impression is that it could be used to improve the code significantly, but adapting it to our use-case and the functions we have available in equistore would take me one or two days and unfortunately I don't have that much time. I will add more comments to explain my current algorithm, which should probably just be taken as a temporary fix to make `to()` scriptable. If, in the future, we want to support gradients of gradients for more operations, we should definitely revisit this topic. Of course, if anybody else in the group (perhaps with a CS background) wants to work on it, that would be an ideal solution